BLD: Allow users to specify BLAS and LAPACK library link order

zerothi opened the pull request:
Prior to this enhancement compiling numpy would forcefully check BLAS/LAPACK
libraries in the following order:

BLAS:
- mkl
- blis
- openblas
- atlas
- accelerate
- NetLIB BLAS

LAPACK:
- mkl
- openblas
- atlas
- accelerate
- NetLIB LAPACK

This is problematic if a user want to build using, say, OpenBLAS but MKL is installed.
Even populating the site.cfg correspondingly one would get a successful build, but
using MKL, if present.

The same applies to OpenBLAS vs. ATLAS etc.

Especially for developers this may be desirable to check performance with various
BLAS/LAPACK libraries.

This fixes the above issues by enabling users to forcefully set the order of loads
via environment variables:

  $> export NUMPY_BLAS_ORDER=openblas,mkl,atlas
  $> python setup.py config ...

would first try OpenBLAS (if existing), then MKL, and finally ATLAS.
In this case the build would fail if neither of OpenBLAS, MKL or ATLAS is present.
I.e. this can also be easierly used to test whether a linking would work. This
is because specifying a single library forces only one library check and has
no fall-back procedure (as requested by the user!).

The same applies to:

 NUMPY_LAPACK_ORDER=openblas,mkl,atlas

This has meant that the `blas_opt_info` and `lapack_opt_info` classes in
`system_info.py` has *completely* changed.

Effectively there is only ONE change:

A fall-back of LAPACK was previously using `get_info('blas')` to get
the BLAS library to correctly link LAPACK. However, this may be undesirable
when the user has OpenBLAS/BLIS/ATLAS in a BLAS only installation but wants
to use the NetLIB LAPACK. Hence now `lapack_opt_info` uses `get_info('blas_opt')`
which does change the fall-back routine slightly. But perhaps for an easier build?

zerothi committed:
ENH: allowed external users to select BLAS and LAPACK library link order

Prior to this enhancement compiling numpy would forcefully check BLAS/LAPACK
libraries in the following order:

BLAS:
- mkl
- blis
- openblas
- atlas
- accelerate
- NetLIB BLAS

- LAPACK
- mkl
- openblas
- atlas
- accelerate
- NetLIB LAPACK

This is problematic if a user want to build using, say, OpenBLAS but MKL is installed.
Even populating the site.cfg correspondingly one would get a successfull build, but
using MKL, if present.

The same applies to OpenBLAS vs. ATLAS etc.

Especially for developers this may be desirable to check performance with various
BLAS/LAPACK libraries.

This fixes the above issues by enabling users to forcefully set the order of loads
via environment variables:

  $> export NUMPY_BLAS_ORDER=openblas,mkl,atlas
  $> python setup.py config ...

would first try OpenBLAS (if existing), then MKL, and finally ATLAS.
In this case the build would fail if neither of OpenBLAS, MKL or ATLAS is present.
I.e. this can also be easierly used to test whether a linking would work. This
is because specifying a single library forces only one library check and has
no fall-back procedure (as requested by the user!).

The same applies to:

 NUMPY_LAPACK_ORDER=openblas,mkl,atlas

This has meant that the blas_opt_info and lapack_opt_info classes in
system_info.py has *completely* changed.

Effectively there is only ONE change:

A fall-back of LAPACK was previously using get_info('blas') to get
the BLAS library to correctly link LAPACK. However, this may be undesirable
when the user has OpenBLAS/BLIS/ATLAS in a BLAS only installation but wants
to use the NetLIB LAPACK. Hence now lapack_opt_info uses get_info('blas_opt')
which does change the fall-back routine slightly. But perhaps for an easier build?

Signed-off-by: Nick Papior <nickpapior@gmail.com>

zerothi commented:
I would be happy to create a test, but it seems extremely build specific and probably difficult in a testing environment?

Do you have a CI just for testing `python setup.py config` commands?

rgommers commented on numpy/distutils/system_info.py:
```diff
@@ -1540,140 +1540,212 @@ def get_atlas_version(**config):
+        # Copy list so we can overwrite
+        lapack_order = self.lapack_order[:]
+
+        user_order = os.environ.get('NUMPY_LAPACK_ORDER', None)
```
better `NPY_LAPACK_ORDER` for consistency with other envvars (`NPY_RELAXED_STRIDES_CHECKING`, `NPY_DISTUTILS_APPEND_FLAGS`, `NPY_PKG_CONFIG_PATH`)

zerothi replied:
Correct! I'll fix!

rgommers commented on numpy/distutils/system_info.py:
```diff
@@ -1540,140 +1540,212 @@ def get_atlas_version(**config):
+                return
 
+        # possible fail-handler? Raise something?
+        # reaching this point means that nothing has been found
```
This should be handled already by the `BlasNotFoundError` and `BlasSrcNotFoundError` usages further up, right? Admittedly those are quite weird. But can't raise anything here, or it will disable `lapack_lite` completely (which may not be a bad thing at some point, but here isn't the place to discuss that).

zerothi replied:
Yeah.

tylerjereddy commented:
> The same applies to OpenBLAS vs. ATLAS etc.

Does it? `ATLAS=None` should do the trick, right?

We can also do `BLAS=None` and `LAPACK=None`. I'm not saying those solve all problems, but curious as to why they aren't discussed in context above re: circumventing backends you don't want to use for building.

rgommers commented:
> I would be happy to create a test, but it seems extremely build specific and probably difficult in a testing environment?

Yeah, it's not really testable on CI. If it passes our existing CI (which uses OpenBLAS and `lapack_lite` IIRC) and some local testing by devs, then it's fine.

> Do you have a CI just for testing python setup.py config commands?

There's a few things in `numpy/distutils/tests/test_system_info.py`, but nothing for BLAS/LAPACK currently. I don't think there's much you could test (easily) there, except perhaps that `NPY_LAPACK_ORDER` is handled correctly.

> Also, sorry I created a commit with ENH, it should have been BLD, I would be happy to change this if so desired. (this PR however has the correct BLD prefix).

No worries. It is ENH anyway arguably, it's a distutils enhancement.

rgommers commented:
The two CI failures are the same - `lapack_lite` build not handled. 

Building with `BLAS=None LAPACK=None ATLAS=None` should still work.

rgommers commented:
Actually, checking that any of `BLAS`, `LAPACK`, `ATLAS` work still when set explicitly would also be good.

zerothi commented:
> > The same applies to OpenBLAS vs. ATLAS etc.
> 
> Does it? `ATLAS=None` should do the trick, right?

Only if MKL is not installed.

> 
> We can also do `BLAS=None` and `LAPACK=None`. I'm not saying those solve all problems, but curious as to why they aren't discussed in context above re: circumventing backends you don't want to use for building.

Agreed, it is probably more meant as a developer thing. And it would also be way easier to have system installed BLAS/LAPACK versions and letting users control this without complicated mixtures of `ATLAS=None MKL=None BLIS=None ...`.

This PR tries to streamline this in a unified way. A second benefit is that the `blas/lapack_opt_info`'s restructure makes it a bit easier to add new libraries.

zerothi commented:
> Actually, checking that any of `BLAS`, `LAPACK`, `ATLAS` work still when set explicitly would also be good.

I'll have another look during next week!

Thanks for both your comments!

zerothi committed:
ENH: amended documentation and changed env-vars as suggested

Also added a test to travis (apparently ATLAS=None... is not tested
on circleCI).

Signed-off-by: Nick Papior <nickpapior@gmail.com>

zerothi committed:
BUG: ensured that warnings/errors are raised when nothing is requested

When a user requests NPY_BLAS/LAPACK_ORDER they can omit Netlib
BLAS/LAPACK. In that case there will not be raised anything.
This commit fixes this issue so that there will always be
issues raised if the user hasn't requested the basic libraries.

Signed-off-by: Nick Papior <nickpapior@gmail.com>

zerothi commented:
@rgommers regarding the `lapack_lite` fail, it seems that this is due to a missing source in `f2c_[cz]_lapack.c` source? Or am I missing something completely? I.e. it seems not to have anything to do with my changes? I can't see why suddenly a missing name `zungqr_` should stem from my changes?

Also do you prefer `NPY_LAPACK` or `NPY_LAPACK_ORDER`? The former could also be used since it is not taken, and 2ndly because the value may be a single value or a comma-separated list?

charris commented:
I see `zungqr_` in `numpy/linalg/lapack_lite/f2c_z_lapack.c`.

zerothi commented:
@charris oh yeah, there it was! :+1: 

I will fix the bug (in my code ;) )!

zerothi committed:
BUG: fixed lapack_lite sources

Signed-off-by: Nick Papior <nickpapior@gmail.com>

rgommers commented:
> Also do you prefer `NPY_LAPACK` or `NPY_LAPACK_ORDER`? The former could also be used since it is not taken, and 2ndly because the value may be a single value or a comma-separated list?

I don't have strong feelings on that one either way. However, does it make sense for it to be a single value given that we already have `LAPACK` for that?

zerothi commented:
Yes and no, if people want the control its main usage will probably be single value uses. Only in special cases can I see that users want to use it multi-value.
Also this value does not reflect any linking step as it is a value equal to the *name* of the library. So perhaps it is fine to have it as it is now.

You'll also note I have amended the build documentation a bit which I think clarifies its use.

zerothi commented:
I think this is ready, I think it is fine to keep the names as is. :)

rgommers commented on numpy/distutils/system_info.py:
```diff
@@ -1539,139 +1539,222 @@ def get_atlas_version(**config):
-            return
+    def _get_info_blas(self):
+        # TODO this was previously blas? Using blas_opt allows NetLIB lapack and BLIS, for instance.
+        #      However, does this break anything!?!?
```
Hmm, this hurts my head. I don't know what the answer is. It looks like the right thing to do as far as I can tell. Let's just say that if it passes CI and all manual testing, it's good to go.

zerothi replied:
ok, good, I will remove the comment.

rgommers commented on numpy/distutils/system_info.py:
```diff
@@ -1539,139 +1539,222 @@ def get_atlas_version(**config):
+        info = self._get_info_lapack()
+        # TODO check whether this is actually necessary to warn about?
+        #      I think the warning is superfluous!
+        # warnings.warn(AtlasNotFoundError.__doc__, stacklevel=2)
```
agree, I think you can delete this - it's a leftover from when ATLAS was our only option I think.

zerothi replied:
agreed

rgommers commented on numpy/distutils/system_info.py:
```diff
@@ -1539,139 +1539,222 @@ def get_atlas_version(**config):
+                elif len(order) > 0:
+                    non_existing.append(order)
+            if len(non_existing) > 0:
+                raise ValueError("lapack_opt_info user defined LAPACK order has unacceptable values: {}".format(non_existing))
```
nitpick: can you break this line to <80 chars?

rgommers commented on numpy/distutils/system_info.py:
```diff
@@ -1539,139 +1539,222 @@ def get_atlas_version(**config):
+    def _calc_info_blas(self):
+        # When one is finally asking for BLAS we need to report
+        # of insufficient usage of optimized BLAS
+        warnings.warn(AtlasNotFoundError.__doc__, stacklevel=3)
```
This is not the right warning anymore. How about just replacing with a string that says something about no optimized BLAS being found?

zerothi replied:
done

rgommers reviewed (commented):
Still need to do some testing, but want to submit my comments already.

zerothi commented on .travis.yml:
```diff
@@ -64,6 +64,13 @@ matrix:
+       - LAPACK=None
+       - ATLAS=None
+       - NPY_BLAS_ORDER=mkl,blis,openblas,atlas,accelerate,blas
+       - NPY_LAPACK_ORDER=MKL,OPENBLAS,ATLAS,ACCELERATE,LAPACK
```
agreed

zerothi committed:
MAINT: fixed last issues and questions according to #13132

Signed-off-by: Nick Papior <nickpapior@gmail.com>

mattip commented on numpy/distutils/system_info.py:
```diff
@@ -1539,139 +1546,219 @@ def get_atlas_version(**config):
-        need_blas = 0
+    def _calc_info_blas(self):
+        # Warn about a non-optimized BLAS library
+        warnings.warn(BlasOptNotFoundError.__doc__, stacklevel=3)
```
This fails when `__doc__` is None

zerothi replied:
I know, but `warnings.warn(*.__doc__, ...)` is used 11 times other than this place in `system_info.py`.

If mine fails, my guess is that using `PYTHONOPTIMIZE=2` is not supported in `numpy` :)

I have now changed with `__doc__ or ''`

mattip replied:
I would imagine using `PYTHONOPTIMIZE` while *building* numpy is even more uncommon than using it while running python. However our CI sets it globally for both build and testing, so the path used for CI must be able to handle it.

The other ones are bugs waiting to be reported/fixed, but not as part of this PR.

rgommers replied:
agreed, it should be robust, but in practice for _building_ it doesn't matter much. people do just pip install and run with `-OO` though, each time we broke that we got a bug report. hence it's in CI now.

zerothi committed:
BUG: fixed PYTHONOPTIMIZE run

Signed-off-by: Nick Papior <nickpapior@gmail.com>

zerothi committed:
MAINT: fixed several PYTHONOPTIMIZE=2 failures

Signed-off-by: Nick Papior <nickpapior@gmail.com>

rgommers reviewed (approved):
I like the `__doc__ or ''` fix, nice and compact.

This all looks good. With `distutils` it's always hard to be really sure, but let's merge it and see how it does. Nice improvement overall.

rgommers commented:
Merged, thanks @zerothi. And thanks @mattip for the review.
